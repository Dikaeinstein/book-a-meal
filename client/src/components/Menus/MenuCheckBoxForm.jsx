import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { ProgressBar } from 'react-materialize';
import Loading from '../util/Loading';
import Label from '../util/Label';
import ConnectedMealsCheckboxList from './MealsCheckboxList';
import Paginate from '../util/Paginate';
import '../../static/sass/materialize.scss';

class MealsCheckBoxForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: this.props.defaultMenuName,
      mealIds: this.props.defaultMenuMealIds,
      checked: this.props.defaultMenuMealIds
        .reduce((obj, key) => ({ ...obj, [`meal-${key}`]: true }), {}),
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleCheck = this.handleCheck.bind(this);
  }

  handleChange(event) {
    this.setState({
      name: event.target.value,
    });
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.handleSubmit({
      name: this.state.name,
      mealIds: this.state.mealIds,
    });
  }

  handleCheck(event) {
    if (event.target.checked) {
      // Extract mealId from input name
      const mealId = event.target.name.slice(5);
      this.setState({
        mealIds: [...this.state.mealIds, parseInt(mealId, 10)],
        checked: {
          ...this.state.checked,
          [event.target.name]: true,
        },
      });
    } else {
      const mealId = event.target.name.slice(5);
      const filteredMealIds = this.state.mealIds.filter(OldMealId =>
        OldMealId !== parseInt(mealId, 10));

      this.setState({
        mealIds: filteredMealIds,
        checked: {
          ...this.state.checked,
          [event.target.name]: false,
        },
      });
    }
  }

  render() {
    return (
      <form
        className="form card"
        onSubmit={this.handleSubmit}
      >
        <h6
          className="text-center"
          style={{ margin: '.5rem 0 1rem 0' }}
        >
          Select meals to add to menu
        </h6>
        <div>
          <Label
            className="label label-block font-weight-bold"
            htmlFor="name"
          >
            Menu Name:
          </Label>
          <input
            name="name"
            placeholder="Enter menu name"
            value={this.state.name}
            onChange={this.handleChange}
          />
        </div>
        <div>
          <h6
            style={{ margin: '.5rem 0' }}
            className="font-weight-bold"
          >
            Meals
          </h6>
          <ConnectedMealsCheckboxList
            meals={this.props.meals}
            checkedMeals={this.state.checked}
            handleCheck={this.handleCheck}
          />
          <Paginate
            onPageChange={this.props.onPageChange}
            nextUrl={this.props.nextUrl}
            previousUrl={this.props.previousUrl}
            style={{ marginTop: '0' }}
          />
        </div>
        <button
          value="Set Menu"
          type="submit"
          className="btn btn-default"
          style={{ margin: '0 auto 1rem auto' }}
        >
          {this.props.action} Menu
        </button>
        {this.props.isSubmitting ?
          <Loading text="setting menu">
            <ProgressBar />
          </Loading>
          : null}
        {!this.props.isSubmitting && this.props.error ?
          <div className="error">{this.props.error}</div>
          : null
        }
      </form>
    );
  }
}

MealsCheckBoxForm.defaultProps = {
  defaultMenuName: 'Menu for Today',
  defaultMenuMealIds: [],
};

MealsCheckBoxForm.propTypes = {
  meals: PropTypes.arrayOf(PropTypes.object).isRequired,
  defaultMenuName: PropTypes.string,
  defaultMenuMealIds: PropTypes.arrayOf(PropTypes.number),
  handleSubmit: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  /* eslint react/require-default-props: 0 */
  error: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.objectOf(PropTypes.string),
  ]),
  action: PropTypes.string.isRequired,
  nextUrl: PropTypes.string,
  previousUrl: PropTypes.string,
  onPageChange: PropTypes.func.isRequired,
};

export default MealsCheckBoxForm;
