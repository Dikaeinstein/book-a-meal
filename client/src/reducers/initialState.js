const initialState = {
  urls: [
    {
      id: 1,
      name: 'Signin',
      link: 'signin',
    },
    {
      id: 2,
      name: 'Signout',
      link: 'signout',
    },
  ],
  menus: {
    isFetching: false,
    data: {
      menu: {
        name: 'Menu for today',
        meals: [],
      },
      menus: [],
    },
    isSet: false,
    isSaving: false,
    fetchError: null,
    saveError: null,
    updateError: null,
    isUpdating: false,
  },
  user: {
    isSubmitting: false,
    loggedIn: false,
    data: {
      id: null,
      name: '',
      email: '',
      role: '',
    },
    error: null,
  },
  orders: {
    isFetching: false,
    isSaving: false,
    isUpdating: false,
    isDeleting: false,
    data: {
      userOrders: [],
      checkedOutOrder: {},
    },
    fetchError: null,
    saveError: null,
    updateError: null,
    deleteError: null,
  },
  meals: {
    isFetching: false,
    isUpdating: false,
    isSaving: false,
    data: [],
    fetchError: null,
    saveError: null,
    updateError: null,
  },
  dashboard: {
    isFetchingTotalOrders: false,
    isFetchingTotalSales: false,
    isUpdating: false,
    isDeleting: false,
    fetchTotalOrdersError: null,
    fetchTotalSalesError: null,
    isFetchingAllOrders: false,
    fetchAllOrdersError: null,
    deleteOrderError: null,
    updateOrderError: null,
    totalAmount: 0,
    totalOrders: 0,
    allOrders: [],
  },
};

export default initialState;
