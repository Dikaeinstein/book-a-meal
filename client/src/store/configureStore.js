import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import throttle from 'lodash.throttle';
import { loadState, saveState } from '../helpers/localStorage';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers';

/**
 * Return redux middleware
 */
const getMiddleware = () => {
  const logger = createLogger();

  const middleware =
    process.env.NODE_ENV !== 'production'
    && process.env.NODE_ENV !== 'test'
      ? applyMiddleware(thunk, logger)
      : applyMiddleware(thunk);

  return middleware;
};

/**
 * Configure the redux store
 *
 * @returns {object} Redux store
 */
const configureStore = () => {
  const persistedState = loadState();

  const state = persistedState
    ? { ...initialState, ...persistedState }
    : initialState;

  const store = createStore(
    rootReducer,
    state,
    compose(
      getMiddleware(),
      window.devToolsExtension ? window.devToolsExtension() : empty => empty,
    ),
  );

  // Throttle the saveState function to be called once every second
  store.subscribe(throttle(() => {
    saveState({ orders: store.getState().orders });
  }, 1000));

  return store;
};

export default configureStore;
