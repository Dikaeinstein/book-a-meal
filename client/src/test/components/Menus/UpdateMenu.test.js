import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import ConnectedUpdateMenu, { UpdateMenu }
  from '../../../components/Menus/UpdateMenu';
import MenuCheckBoxForm from '../../../components/Menus/MenuCheckBoxForm';
import initialState from '../../../reducers/initialState';

describe('<UpdateMenu />', () => {
  const meals = [
    { id: 1, name: 'test meal', price: 4000 },
    { id: 2, name: 'test meal', price: 4000 },
  ];
  const props = {
    setMenu: jest.fn(),
    error: null,
    menu: { name: 'Test menu for today', meals },
    fetchMeals: jest.fn(),
    isUpdating: false,
    isFetching: false,
    menuId: 1,
    meals,
    modifyMenu: jest.fn(),
    closeModal: jest.fn(),
    currentUrl: 'http://testurl',
  };
  it('should render with correct props', () => {
    const mockStore = configureStore();
    const store = mockStore(initialState);
    const wrapper = shallow(<ConnectedUpdateMenu {...props} store={store} />);
    expect(wrapper.props().error).toBeNull();
    expect(wrapper.props().isUpdating).toBeFalsy();
    expect(wrapper.props().meals.length).toEqual(0);
  });
  it('should call handleSubmit when menu is set', () => {
    const spy = jest.spyOn(UpdateMenu.prototype, 'handleSubmit');
    const wrapper = shallow(<UpdateMenu {...props} />);
    const updateMenuForm = wrapper.find(MenuCheckBoxForm);
    updateMenuForm.props().handleSubmit({ mealIds: [1, 2] });
    expect(spy).toHaveBeenCalled();
  });
  it('should match snapshot', () => {
    const tree = shallow(<UpdateMenu {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
