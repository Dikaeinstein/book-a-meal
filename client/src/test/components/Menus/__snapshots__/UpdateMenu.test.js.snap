// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdateMenu /> should match snapshot 1`] = `
<section>
  <h5
    className="text-center"
  >
    Update Menu
  </h5>
  <MealsCheckBoxForm
    action="Set"
    currentUrl="http://testurl"
    defaultMenuMealIds={
      Array [
        1,
        2,
      ]
    }
    defaultMenuName="Test menu for today"
    error={null}
    handleSubmit={[Function]}
    isSubmitting={false}
    meals={
      Array [
        Object {
          "id": 1,
          "name": "test meal",
          "price": 4000,
        },
        Object {
          "id": 2,
          "name": "test meal",
          "price": 4000,
        },
      ]
    }
    onPageChange={[Function]}
  />
</section>
`;
